aws-storage-s3/src/test/java/com/amplifyframework/storage/s3/StorageComponentTest.java

```diff
@@ -0,0 +1,235 @@
+
+    private StorageCategory storage;
+    private StorageService storageService;
+
+    /**
+     * Sets up Storage category by registering a mock AWSS3StoragePlugin
+     * instance to Amplify and configuring.
+     * @throws AmplifyException if Amplify fails to configure with mock
+     *         Storage category configuration.
+     */
+    @Before
+    public void setup() throws AmplifyException {
+        this.storage = new StorageCategory();
+        this.storageService = mock(StorageService.class);
+        StorageService.Factory storageServiceFactory = (context, region, bucket) -> storageService;
+        IdentityIdProvider identityIdProvider = RandomString::string;
+        this.storage.addPlugin(new AWSS3StoragePlugin(storageServiceFactory, identityIdProvider));
+        this.storage.configure(buildConfiguration(), mock(Context.class));
+    }
+
+    private static StorageCategoryConfiguration buildConfiguration() {
+        StorageCategoryConfiguration configuration = new StorageCategoryConfiguration();
+        JSONObject storageJson;
+        try {
+            storageJson = new JSONObject()
+                .put("plugins", new JSONObject()
+                    .put("awsS3StoragePlugin", new JSONObject()
+                        .put("region", "us-east-1")
+                        .put("bucket", "hamburger-bucket")));
+            configuration.populateFromJSON(storageJson);
+        } catch (JSONException jsonException) {
+            throw new RuntimeException(jsonException);
+        }
+        return configuration;
+    }
+
+    /**
+     * Test that calling download file method from Storage category correctly
+     * invokes the registered AWSS3StoragePlugin instance and returns a
+     * {@link StorageDownloadFileResult} with correct file path.
+     * @throws StorageException when an error is encountered while downloading
+     */
+    @Test
+    public void testDownloadToFileGetsFile() throws StorageException {
+        final String fromRemoteKey = RandomString.string();
+        final String toLocalPath = RandomString.string();
+
+        // Since we use a mock StorageService, it will return a null
+        // result by default. We need a non-null transfer observer.
+        // One option is to mock that, too.
+        TransferObserver observer = mock(TransferObserver.class);
+        when(storageService.downloadToFile(anyString(), any(File.class)))
+                .thenReturn(observer);
+
+        // Since we use a mock TransferObserver, it has no internal logic
+        // to know to call back the listener! So, we simulate the success
+        // callback, as part of our "happy path" test.
+        doAnswer(invocation -> {
+            TransferListener listener = invocation.getArgument(0);
+            listener.onStateChanged(0, TransferState.COMPLETED);
+            return null;
+        })
+                .when(observer)
+                .setTransferListener(any(TransferListener.class));
```

Comment:
You could probably pull this out to a helper method, like:
```
mockTransferListenerResult(TransferState.COMPLETED);

private void mockTransferListenerResult(TransferState transferState) {
    doAnswer(invocation -> {
        TransferListener listener = invocation.getArgument(0);
        listener.onStateChanged(0, transferState);
        return null;
    }).when(observer)
      .setTransferListener(any(TransferListener.class));
}
```

It'd be cool to add a few more tests that test what happens when the transfer state is `FAILED`, or whatever -- other states than `COMPLETE`.

Reply:
Storage operations, as currently implemented, does not handle other transfer states. Not sure if that's by design.